azkaban-common/src/main/java/azkaban/scheduler/MissedSchedulesManager.java

```diff
@@ -0,0 +1,199 @@
+    if (this.missedSchedulesExecutor != null) {
+      this.missedSchedulesExecutor.shutdown();
+      if (!this.missedSchedulesExecutor.awaitTermination(30, TimeUnit.SECONDS)) {
+        this.missedSchedulesExecutor.shutdownNow();
+      }
+    }
+    return true;
+  }
+
+  /**
+   * A MissedSchedule Task is a unit that store the critical information to execute a missedSchedule operation.
+   * A Task is capable to
+   * 1. notice user
+   * 2. execute back execution if enabled feature flag (TODO)
+   * when a missed schedule happens.
+   * */
+  public static class MissedSchedulesOperationTask implements Runnable {
+    private static final Logger LOG = LoggerFactory.getLogger(MissedSchedulesOperationTask.class);
+    private final Emailer emailer;
+    // emailRecipients to notify users the schedules missed
+    private final List<String> emailRecipients;
+    // the email body
+    private final String emailMessage;
+    // back execute action
+    // if user disabled the config, the action could be null.
+    private ExecuteFlowAction executeFlowAction;
+
+    public MissedSchedulesOperationTask(List<Long> missedScheduleTimes,
+        Emailer emailer,
+        List<String> emailRecipients,
+        boolean backExecutionEnabled,
+        ExecuteFlowAction executeFlowAction) {
+      this.emailer = emailer;
+      this.emailRecipients = emailRecipients;
+      this.executeFlowAction = backExecutionEnabled ? executeFlowAction : null;
+      final String missScheduleTimestampToDate = formatTimestamps(missedScheduleTimes);
+      MessageFormat messageFormat = new MessageFormat("This is autogenerated email to notify to flow owners that"
+          + " the flow {0} in project {1} has scheduled the executions on {2} but failed to trigger on time. "
+          + (backExecutionEnabled ? "Back execution will start soon as you enabled the config." : ""));
+
+      this.emailMessage = messageFormat.format(
+          new Object[]{executeFlowAction.getFlowName(), executeFlowAction.getProjectName(), missScheduleTimestampToDate});
+    }
+
+    private String formatTimestamps(List<Long> missScheduleTimestamps) {
+      final SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
+      List<String> datesFromTimestamps = missScheduleTimestamps.stream()
+          .map(timestamps -> sdf.format(new Date(timestamps)))
+          .collect(Collectors.toList());
+      return String.join(",", datesFromTimestamps);
+    }
+
+    @Override
+    public void run() {
+      try {
+        this.emailer.sendEmail(this.emailRecipients, "Missed Azkaban Schedule Notification", this.emailMessage);
+        if (this.executeFlowAction != null) {
+          this.executeFlowAction.doAction();
+        }
+      } catch (InterruptedException e) {
+        String warningMsg = "MissedScheduleTask thread is being interrupted, throwing out the exception";
+        LOG.warn(warningMsg, e);
+        throw new RuntimeException(warningMsg, e);
```

Comment:
Should we throw an AzkabanException here?

Reply:
Sure